ddtrace/internal/_encoding.pyx

```diff
@@ -0,0 +1,357 @@
+                    if ret != 0: return ret
+                    ret = self._pack(o.start_ns)
+                    if ret != 0: return ret
+
+                    ret = self._pack_bytes(<char *>b"duration")
+                    if ret != 0: return ret
+                    ret = self._pack(o.duration_ns)
+                    if ret != 0: return ret
+
+                    if has_span_type:
+                        ret = self._pack_bytes(<char *>b"type")
+                        if ret != 0: return ret
+                        ret = self._pack(o.span_type)
+                        if ret != 0: return ret
+
+                    if has_meta:
+                        ret = self._pack_bytes(<char *>b"meta")
+                        if ret != 0: return ret
+                        ret = self._pack(o.meta)
+                        if ret != 0: return ret
+
+                    if has_metrics:
+                        ret = self._pack_bytes(<char *>b"metrics")
+                        if ret != 0: return ret
+                        ret = self._pack(o.metrics)
+                        if ret != 0: return ret
+            else:
+                PyErr_Format(TypeError, b"can not serialize '%.200s' object", Py_TYPE(o).tp_name)
+            return ret
+
+    cdef int _pack_bytes(self, char *rawval):
+        cdef int ret
+        cdef dict d
+        cdef Py_ssize_t L
+        L = len(rawval)
+        if L > ITEM_LIMIT:
+            PyErr_Format(ValueError, b"%.200s object is too large", Py_TYPE(rawval).tp_name)
+        ret = msgpack_pack_bin(&self.pk, L)
+        if ret == 0:
+            ret = msgpack_pack_raw_body(&self.pk, rawval, L)
+        return ret
+
+    cpdef pack(self, object obj):
+        cdef int ret
+        try:
+            ret = self._pack(obj)
+        except:
+            self.pk.length = 0
+            raise
+        if ret:  # should not happen.
+            raise RuntimeError("internal error")
+
+        if self.autoreset:
+            buf = PyBytes_FromStringAndSize(self.pk.buf, self.pk.length)
+            self.pk.length = 0
+            return buf
+
+    def reset(self):
+        """Reset internal buffer.
+
+        This method is useful only when autoreset=False.
+        """
+        self.pk.length = 0
+
+    def bytes(self):
+        """Return internal buffer contents as bytes object"""
+        return PyBytes_FromStringAndSize(self.pk.buf, self.pk.length)
+
+    def getbuffer(self):
+        """Return view of internal buffer."""
+        return buff_to_buff(self.pk.buf, self.pk.length)
+
+
+cdef class TraceMsgPackEncoder(object):
+    content_type = "application/msgpack"
+
+    cpdef decode(self, data):
+        import msgpack
+        if msgpack.version[:2] < (0, 6):
+            return msgpack.unpackb(data)
+        return msgpack.unpackb(data, raw=True)
+
+    cpdef encode_trace(self, trace):
+        return Packer().pack(trace)
+
+    cpdef encode_traces(self, traces):
```

Comment:
we can do type hinting on all of these `cpdef` for return types and arguments, might help optimize things a bit.

Reply:
We don't gain much of anything here since they're pyobjects in and out. I can try type hinting the argument as `list` and see if we get anything out of it.